SMT2 back-end: flatten with_exprt operands

tautschnig opened the pull request:
Solvers without datatypes support require flattening towards bitvectors. This is also true for array-typed expressions when the array is not at the top level (despite possible array-theory support). This changes makes sure we will flatten such array-typed expressions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
SMT2 back-end: flatten with_exprt operands

Solvers without datatypes support require flattening towards bitvectors.
This is also true for array-typed expressions when the array is not at
the top level (despite possible array-theory support). This changes
makes sure we will flatten such array-typed expressions.

tautschnig commented:
~~Note to reviewers: #8637 includes a test set-up with tests passing once the changes in this PR are applied. I am working on extracting a minified variant thereof to include as regression test.~~ Edit: done.

tautschnig committed:
CI: Make smt2_solver available to all regression tests

We will want to use the solver for more than just selected CBMC
regressions.

rod-chapman commented:
This branch does seems to fix the array flattening issue on the mlda-native code when using bitwuzla.

TWO functions still fail with a different error though. For example, the unpack-hints function in mldsa-native:
```
cbmc --bitwuzla --outfile b.smt2 gotos/unpack_hints_harness.goto
bitwuzla b.smt2
```
yields
```
[error] b.smt2:18036:529: expected terms of same sort at indices 1 and 2 as argument to 'ite'
```
Is this related, or a completely different problem?

tautschnig commented:
Thank you for testing! I'll take a look tomorrow morning to figure out whether the remaining problem (besides `unpack_hints`, which is the other function that still is having issues?) is unrelated or the proposed bugfix is incomplete.

rod-chapman commented:
Results: testing with CBMC (this branch), Z4 4.12.6 and Bitwuzla 0.7.0 on the mldsa-native, HEAD of main branch - all 134 functions/proofs OK.

tautschnig committed:
SMT2 back-end: ite-conversion must uphold type consistency

WITH expressions in an ite expression could previously trigger
one-sided expression flattening.

rod-chapman commented:
Testing this branch (with 3rd commit) on mlkem-native main/HEAD - all 126 proofs OK.

rod-chapman commented:
Testing this branch (with 3rd commit) on mldsa-native - 134 proofs all OK, and a bit faster too.
